From: Tommy Huang <tommy_huang@aspeedtech.com>
Date: Mon, 9 Mar 2026 08:44:45 +0800
Subject: [PATCH 6.12 1/1] i2c: ast2600: arrange the transfer judgement rule

Upstream-Status: Backport [AspeedTech-BMC/linux aspeed-master-v6.12 c737310e59ba6ae30ae060a4c92f0995f096f90b, SDK v00.07.03]

Arrange the i2c transfer judgement rule.

Signed-off-by: Tommy Huang <tommy_huang@aspeedtech.com>
Change-Id: I81fe58b6acbb5dd210b561d924b770f82182ed4f
(cherry picked from commit d6c8e38d7d5f7561f96a87128fec55fb8b806d86)
---
 drivers/i2c/busses/i2c-ast2600.c | 53 +++++++++++++++-----------------
 1 file changed, 25 insertions(+), 28 deletions(-)

diff --git a/drivers/i2c/busses/i2c-ast2600.c b/drivers/i2c/busses/i2c-ast2600.c
index 8da094dcd..f27aff994 100644
--- a/drivers/i2c/busses/i2c-ast2600.c
+++ b/drivers/i2c/busses/i2c-ast2600.c
@@ -2906,10 +2906,26 @@ static int ast2600_i2c_probe(struct platform_device *pdev)
 		i2c_bus->ac_timing_config = ast2700_i2c_ac_timing_config;
 		i2c_bus->irq_err_to_errno = ast2700_i2c_irq_err_to_errno;
 		i2c_bus->irq_clear = ast2700_i2c_irq_clear;
+
+		/* AST2700: default DMA, allow override */
+		i2c_bus->mode = DMA_MODE;
 	} else {
 		i2c_bus->ac_timing_config = ast2600_i2c_ac_timing_config;
 		i2c_bus->irq_err_to_errno = ast2600_i2c_irq_err_to_errno;
 		i2c_bus->irq_clear = ast2600_i2c_irq_clear;
+
+		/* AST2600: default BUFF, allow override */
+		i2c_bus->mode = BUFF_MODE;
+	}
+
+	/* override the transfer mode */
+	if (!device_property_read_string(dev, "aspeed,transfer-mode", &xfer_mode)) {
+		if (!strcmp(xfer_mode, "dma"))
+			i2c_bus->mode = DMA_MODE;
+		else if (!strcmp(xfer_mode, "byte"))
+			i2c_bus->mode = BYTE_MODE;
+		else
+			i2c_bus->mode = BUFF_MODE;
 	}
 
 #if IS_ENABLED(CONFIG_I2C_SLAVE)
@@ -2922,42 +2938,23 @@ static int ast2600_i2c_probe(struct platform_device *pdev)
 	i2c_bus->multi_master = device_property_read_bool(dev, "multi-master");
 
 	if (i2c_bus->version == AST2700) {
-		/* AST2700: default DMA, allow override */
-		i2c_bus->mode = DMA_MODE;
-
-		if (!device_property_read_string(dev, "aspeed,transfer-mode", &xfer_mode)) {
-			if (!strcmp(xfer_mode, "dma"))
-				i2c_bus->mode = DMA_MODE;
-			else if (!strcmp(xfer_mode, "byte"))
-				i2c_bus->mode = DMA_MODE;
-			else
-				i2c_bus->mode = BUFF_MODE;
-		}
+		/* AST2700 would not allow byte mode */
+		if (i2c_bus->mode == BYTE_MODE)
+			i2c_bus->mode = DMA_MODE;
 
 		if (i2c_bus->mode == BUFF_MODE) {
 			i2c_bus->buf_base = devm_platform_get_and_ioremap_resource(pdev, 1, &res);
-			if (IS_ERR(i2c_bus->buf_base))
+			if (IS_ERR(i2c_bus->buf_base)) {
 				i2c_bus->mode = DMA_MODE;
-			else
+			} else {
 				i2c_bus->buf_size = resource_size(res) / 4;
 
-			/* ast2700 cfg disable dma */
-			writel(readl(i2c_bus->reg_base + I2CC_VERSION_CTRL) & ~FUNC_CFG_DMA_EN,
-				   i2c_bus->reg_base + I2CC_VERSION_CTRL);
+				/* ast2700 cfg disable dma */
+				writel(readl(i2c_bus->reg_base + I2CC_VERSION_CTRL) & ~FUNC_CFG_DMA_EN,
+					   i2c_bus->reg_base + I2CC_VERSION_CTRL);
+			}
 		}
 	} else {
-		/* AST2600: default BUFF, allow override */
-		i2c_bus->mode = BUFF_MODE;
-
-		if (!device_property_read_string(dev, "aspeed,transfer-mode", &xfer_mode)) {
-			if (!strcmp(xfer_mode, "dma"))
-				i2c_bus->mode = DMA_MODE;
-			else if (!strcmp(xfer_mode, "byte"))
-				i2c_bus->mode = BYTE_MODE;
-			else
-				i2c_bus->mode = BUFF_MODE;
-		}
-
 		if (i2c_bus->mode == BUFF_MODE) {
 			i2c_bus->buf_base = devm_platform_get_and_ioremap_resource(pdev, 1, &res);
 			if (IS_ERR(i2c_bus->buf_base))
